File: tests/test_animation.py
# This file is part of GreenMoon2D, a 2D game engine for Python
# written by Willi Kappler, MIT license.
#
# See: https://github.com/willi-kappler/green_moon_2d

import unittest
import time

from green_moon_2d.gm_animation import GMAnimType, GMAnimation


class TestAnimation(unittest.TestCase):
    def test_empty_frames(self):
        """
        Test for exception if user provides empty frames.
        """

        with self.assertRaises(AssertionError):
            a1 = GMAnimation([])
            a1.current_frame = 0

    def test_change_type(self):
        """
        Test changing the type of the animation.
        """

        frames = [(1, 100), (2, 100), (3, 200)]
        last_frame = len(frames) - 1
        a1 = GMAnimation(frames)

        self.assertEqual(a1.anim_type, GMAnimType.FORWARD)

        a1.current_frame = 1
        a1.change_type(GMAnimType.FORWARD)
        self.assertEqual(a1.current_frame, 0)

        a1.current_frame = 1
        a1.change_type(GMAnimType.BACKWARD)
        self.assertEqual(a1.current_frame, last_frame)

        a1.current_frame = 1
        a1.change_type(GMAnimType.FORWARD_LOOP)
        self.assertEqual(a1.current_frame, 0)

        a1.current_frame = 1
        a1.change_type(GMAnimType.BACKWARD_LOOP)
        self.assertEqual(a1.current_frame, last_frame)

        a1.current_frame = 1
        a1.change_type(GMAnimType.PINGPONG_F)
        self.assertEqual(a1.current_frame, 0)

        a1.current_frame = 1
        a1.change_type(GMAnimType.PINGPONG_B)
        self.assertEqual(a1.current_frame, last_frame)

    def test_get_frame_index(self):
        """
        Test returning the correct frame index.
        """

        a1 = GMAnimation([(5, 100), (10, 100), (8, 300)])
        self.assertEqual(a1.get_frame_index(), 5)

        a1.current_frame = 1
        self.assertEqual(a1.get_frame_index(), 10)

        a1.current_frame = 2
        self.assertEqual(a1.get_frame_index(), 8)

    def test_set_timer_duration(self):
        """
        Test setting the correct timer duration.
        """

        a1 = GMAnimation([(5, 100), (10, 550), (8, 310)])
        a1.set_timer_duration()
        self.assertEqual(a1.timer.duration, 100)

        a1.current_frame = 1
        a1.set_timer_duration()
        self.assertEqual(a1.timer.duration, 550)

        a1.current_frame = 2
        a1.set_timer_duration()
        self.assertEqual(a1.timer.duration, 310)

    def test_update1(self):
        """
        Test updating the animation with FORWARD type.
        """

        a1 = GMAnimation([(2, 100), (4, 200), (7, 100), (1, 200)])
        a1.update()
        self.assertEqual(a1.current_frame, 0)
        self.assertEqual(a1.get_frame_index(), 2)

        time.sleep(0.11)
        a1.update()
        self.assertEqual(a1.current_frame, 1)
        self.assertEqual(a1.get_frame_index(), 4)

        time.sleep(0.11)
        a1.update()
        self.assertEqual(a1.current_frame, 1)
        self.assertEqual(a1.get_frame_index(), 4)

        time.sleep(0.10)
        a1.update()
        self.assertEqual(a1.current_frame, 2)
        self.assertEqual(a1.get_frame_index(), 7)

        time.sleep(0.11)
        a1.update()
        self.assertEqual(a1.current_frame, 3)
        self.assertEqual(a1.get_frame_index(), 1)

        time.sleep(0.21)
        a1.update()
        self.assertEqual(a1.current_frame, 3)
        self.assertEqual(a1.get_frame_index(), 1)
        self.assertEqual(a1.active, False)


if __name__ == "__main__":
    unittest.main()

File: tests/test_animation.cpp
/*
    GreenMoon2D
    SPDX-License-Identifier: MIT
    Written by Willi Kappler, MIT License
    https://github.com/willi-kappler/green_moon_2d

    This file contains the tests for the animation class.

    Run only configuration tests:
    xmake run -w ./ gm2d_test [animation]
*/

// STD includes:
#include <thread>

// External includes:
#include <snitch/snitch.hpp>

// Local includes:
#include "gm2d/gm_animation.hpp"

using namespace gm2d;

TEST_CASE("Create animation.", "[animation]") {
    GMAnimation a1{{5, 100}, {10, 100}, {8, 300}};
    REQUIRE(a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    std::vector<std::pair<uint16_t, uint16_t>> frames{{5, 100}, {10, 100}, {8, 300}};
    GMAnimation a2(frames);
    REQUIRE(a2.gm_finished());
    REQUIRE(a2.gm_get_frame_index() == 5);
}

TEST_CASE("Change animation repetition.", "[animation]") {
    GMAnimation a1{{5, 100}, {10, 100}, {8, 300}};
    a1.gm_change_repetition(GMRepetition::FIXED);
    REQUIRE(a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    a1.gm_change_repetition(GMRepetition::FORWARD);
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    a1.gm_change_repetition(GMRepetition::BACKWARD);
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);

    a1.gm_change_repetition(GMRepetition::FORWARD_LOOP);
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    a1.gm_change_repetition(GMRepetition::BACKWARD_LOOP);
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);

    a1.gm_change_repetition(GMRepetition::PINGPONG_F);
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    a1.gm_change_repetition(GMRepetition::PINGPONG_B);
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);
}

TEST_CASE("Test update with fixed repetition.", "[animation]") {
    GMAnimation a1{{5, 100}, {10, 100}, {8, 300}};
    a1.gm_change_repetition(GMRepetition::FIXED);
    REQUIRE(a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    auto const sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);
}

TEST_CASE("Test update with forward repetition.", "[animation]") {
    GMAnimation a1{{5, 100}, {10, 100}, {8, 300}};
    a1.gm_change_repetition(GMRepetition::FORWARD);
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    auto const sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 10);

    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);

    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);
}

TEST_CASE("Test update with backward repetition.", "[animation]") {
    GMAnimation a1{{5, 100}, {10, 100}, {8, 300}};
    a1.gm_change_repetition(GMRepetition::BACKWARD);
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);

    auto sleep_time = std::chrono::milliseconds(300);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 10);

    sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);
}

TEST_CASE("Test update with forward loop repetition.", "[animation]") {
    GMAnimation a1{{5, 100}, {10, 100}, {8, 300}};
    a1.gm_change_repetition(GMRepetition::FORWARD_LOOP);
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    auto sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 10);

    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);

    sleep_time = std::chrono::milliseconds(300);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 10);
}

TEST_CASE("Test update with backward loop repetition.", "[animation]") {
    GMAnimation a1{{5, 100}, {10, 100}, {8, 300}};
    a1.gm_change_repetition(GMRepetition::BACKWARD_LOOP);
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);

    auto sleep_time = std::chrono::milliseconds(300);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 10);

    sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);

    sleep_time = std::chrono::milliseconds(300);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 10);
}

TEST_CASE("Test update with pingpong forward repetition.", "[animation]") {
    GMAnimation a1{{5, 100}, {10, 100}, {8, 300}};
    a1.gm_change_repetition(GMRepetition::PINGPONG_F);
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    auto sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 10);

    sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);

    sleep_time = std::chrono::milliseconds(300);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 10);

    sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 10);

    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);
}

TEST_CASE("Test update with pingpong backward repetition.", "[animation]") {
    GMAnimation a1{{5, 100}, {10, 100}, {8, 300}};
    a1.gm_change_repetition(GMRepetition::PINGPONG_B);
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);

    auto sleep_time = std::chrono::milliseconds(300);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 10);

    sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);

    sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 10);

    sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 8);

    sleep_time = std::chrono::milliseconds(300);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 10);

    sleep_time = std::chrono::milliseconds(100);
    std::this_thread::sleep_for(sleep_time);
    a1.gm_update();
    REQUIRE(!a1.gm_finished());
    REQUIRE(a1.gm_get_frame_index() == 5);
}
